lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -0,0 +1,1022 @@
+  friend class swift::ParseableInterfaceModuleLoader;
+  ASTContext &ctx;
+  llvm::vfs::FileSystem &fs;
+  DiagnosticEngine &diags;
+  const StringRef modulePath;
+  const std::string interfacePath;
+  const StringRef moduleName;
+  const StringRef prebuiltCacheDir;
+  const StringRef cacheDir;
+  const SourceLoc diagnosticLoc;
+  DependencyTracker *const dependencyTracker;
+  const ModuleLoadingMode loadMode;
+
+  ParseableInterfaceModuleLoaderImpl(
+    ASTContext &ctx, StringRef modulePath, StringRef interfacePath,
+    StringRef moduleName, StringRef cacheDir, StringRef prebuiltCacheDir,
+    SourceLoc diagLoc, DependencyTracker *dependencyTracker = nullptr,
+    ModuleLoadingMode loadMode = ModuleLoadingMode::PreferSerialized)
+  : ctx(ctx), fs(*ctx.SourceMgr.getFileSystem()), diags(ctx.Diags),
+    modulePath(modulePath), interfacePath(interfacePath),
+    moduleName(moduleName), prebuiltCacheDir(prebuiltCacheDir),
+    cacheDir(cacheDir), diagnosticLoc(diagLoc),
+    dependencyTracker(dependencyTracker), loadMode(loadMode) {}
+
+  /// Construct a cache key for the .swiftmodule being generated. There is a
+  /// balance to be struck here between things that go in the cache key and
+  /// things that go in the "up to date" check of the cache entry. We want to
+  /// avoid fighting over a single cache entry too much when (say) running
+  /// different compiler versions on the same machine or different inputs
+  /// that happen to have the same short module name, so we will disambiguate
+  /// those in the key. But we want to invalidate and rebuild a cache entry
+  /// -- rather than making a new one and potentially filling up the cache
+  /// with dead entries -- when other factors change, such as the contents of
+  /// the .swiftinterface input or its dependencies.
+  std::string getCacheHash(const CompilerInvocation &SubInvocation) {
+    // Start with the compiler version (which will be either tag names or revs).
+    // Explicitly don't pass in the "effective" language version -- this would
+    // mean modules built in different -swift-version modes would rebuild their
+    // dependencies.
+    llvm::hash_code H = hash_value(swift::version::getSwiftFullVersion());
+
+    // Simplest representation of input "identity" (not content) is just a
+    // pathname, and probably all we can get from the VFS in this regard
+    // anyways.
+    H = hash_combine(H, interfacePath);
+
+    // Include the target CPU architecture. In practice, .swiftinterface files
+    // will be in architecture-specific subdirectories and would have
+    // architecture-specific pieces #if'd out. However, it doesn't hurt to
+    // include it, and it guards against mistakenly reusing cached modules
+    // across architectures.
+    H = hash_combine(H, SubInvocation.getLangOptions().Target.getArchName());
+
+    // The SDK path is going to affect how this module is imported, so include
+    // it.
+    H = hash_combine(H, SubInvocation.getSDKPath());
```

Comment:
Is this a canonicalized path? If not, maybe it should be?

Reply:
Looks like the only canonicalization we perform is removing a trailing '/' when it's set. Given that this is a command-line flag, we probably should.